### ISSUE TITLE
PIM libraries

### ISSUE THREAD

**Issue by matthiasbeyer** [ErichDonGubler/not-yet-awesome-rust#26]
If I understand correctly, we are encouraged to propose new entries in the README? If yes, ... here we go:

## Mail

* There's [lettre_email](https://crates.io/crates/lettre_email) for _building_ emails, but there's no high-level library for _parsing_ email (The [email](https://niax.github.io/rust-email/email/) crate is rather low-level IMO).

* There's no library for transparently handling Maildir and/or mbox

## Contacts

* There's [vobject](https://crates.io/crates/vobject) which handles contact files (vcard). There are several others, but after evaluating it seems that vobject is the best one. Though it lacks a nice high-level interface (I contributed the foundation for a nice high-level interface, but did not have enough time for polishing yet).

## Calendar

* The [vobject](https://crates.io/crates/vobject) crate also handles calendar stuff, though the same as above applies: The foundations for a high-level interface are there, but it is not polished yet.

## Ledger

* This is a bit more niche: I would love to see a parser and high-level interface for [beancount syntax/data](http://furius.ca/beancount/)

## Parser for Media-formats

(I did not do a lot of research on this one, TBH)

* A Parser and Interface for reading/writing MP3 metadata (or any audio format)
* A Parser and Interface for reading/writing Movie format metadata

## Weather data / news

(I did not do a lot of research on this one, TBH)

* A crate for reading weather data, possibly from different sources
* A crate for reading news data. I guess this is mostly done via RSS, so a high-level RSS reader would be awesome (as library)

# The larger goal

All my proposals here are focused (in one way or another) on **P**ersonal **I**nformation **M**anagement (PIM). That's because I develop a PIM suite for the commandline (but extensible to TUI/GUI/WUI). If anyone wants to develop (for example) a ledger, todo-tracker or workout-tracker, they are **more than welcome** to contribute, so we can join forces. The framework which I developed in the last two years is rather stable and **can be used today to develop PIM tools**!
Just talk to me: [https://imag-pim.org](https://imag-pim.org) / [https://github.com/matthiasbeyer/imag](https://github.com/matthiasbeyer/imag)

---

I'll file an PR to update the list, if you want!

**Comment by ErichDonGubler** [comment]
Hey! Thanks for taking the time to make some suggestions! :) I haven't had time yet to look in significant detail at your proposals, but it looks like there will be some good discussion here!

**Comment by ErichDonGubler** [comment]
## Contacts/Calendar

Have you evaluated the [`ical`](https://github.com/Peltoche/ical-rs) crate or the supporting libraries for [`imag`](https://github.com/matthiasbeyer/imag)? They seem to have more tools a "high-level" crate would need, though I'm not sure what exactly you'd want. What do you want beyond parsing and emitting vCard/iCal?

## Ledger

Again, what high-level APIs are missing here? What would you expect? Would you want something with parity to the original `beancount` project?

## Weather

In the interest of providing concrete action items, what services would you suggest? Is there an API reference anywhere? I've found these:

* [`weather`](https://crates.io/crates/weather): Developed for a binary; it seems very WIP and that it only works for Weather Underground ATM.
* [`ghcn-daily`](https://crates.io/crates/ghcn-daily) -- a parser for historical data collected at the service of the same name. Probably not useful for your use case, though the concept itself sounds interesting.
* [`darksky`](https://docs.rs/darksky/0.8.0/darksky/) and [`forecast`](https://jgrillo.github.io/forecast-rs/forecast/index.html) use [DarkSky](https://darksky.net/).

I'd say the use case currently not covered is aggregating data from multiple APIs, but I don't know if there's a standard interface that is in common between these different services or others that haven't been encapsulated with Rust bindings yet. Otherwise, this doesn't seem to have enough concrete detail to be actionable yet (perhaps we should opt for specific services instead?), but I'm definitely no expert in this domain. 

## News

* Looks like [`rss`](https://docs.rs/rss/1.2.1/rss/) is pretty much covered from the docs I see. :)

---

I would love help getting [this PR](https://github.com/ErichDonGubler/not-yet-awesome-rust/pull/31) finished based on discussion we have/will have here -- especially when looking for details I may have missed (which I'm sure I did...). Let me know what you think!

**Comment by matthiasbeyer** [comment]
@ErichDonGubler `ical` is over a year old and has only a low-level interface to the properties. `vobject` (which I contributed to) has a better interface and allows building objects as well ... and since my contribution there's also a high-level interface (which is not as nice as I would like it to be, but at least it is _there_)!

I'm not sure why you refer to imag? I need the vcard/icalendar crates for writing imag tools... 

---

> Again, what high-level APIs are missing here? What would you expect? Would you want something with parity to the original beancount project?

I'd love to have a crate which can parse the beancount data format and allows reading and writing beancount expressions from Rust code.

Developing my own ledger for imag would be a hell of a job and beancount is there already... so integrating it seems natural (as we did with taskwarrior).

---

Weather: TIL, thanks for providing links!
News: Ah, right! Awesome!

---

Will comment in the PR.
